### PR TITLE
[develop] Adding additional requisites to state compiler

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Adding four additional requisites, ``require_any`` ``watch_any`` ``onchanges_any`` ``onfail_any``, to the state compiler.

### What issues does this PR fix or reference?
N/A

### Previous Behavior
Previously these requisites did not exist.

### New Behavior
Adding four additional requisites, ``require_any`` ``watch_any`` ``onchanges_any`` ``onfail_any``, to the state compiler.

### Tests written?
Yes

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: doc/ref/states/requisites.rst [hunk 1]
```diff
@@ -245,6 +252,44 @@ This will add all of the state declarations found in the given sls file. This me
...
+The use of ``require_any`` demands that one of the required state executes before the
+dependent state. The state containing the ``require_any`` requisite is defined as the
+dependent state. The states specified in the ``require_any`` statement are defined as the
+required states. If at least one of the required state's execution succeeds, the dependent state
+will then execute. If at least one of the required state's execution fails, the dependent state
+will not execute.
+
+.. code-block:: yaml
+
+	A:
```
Comment (thatch45) [inline doc/ref/states/requisites.rst]: are these tabs?!?!?!?!?

File: doc/ref/states/requisites.rst [hunk 2]
```diff
@@ -482,6 +608,37 @@ if any of the watched states changes.
...
+
+.. versionadded:: Oxygen
+
+The ``onchanges_any`` requisite makes a state only apply one of the required states
+generates changes, and if one of the watched state's "result" is ``True``. This can be
+a useful way to execute a post hook after changing aspects of a system.
+
+.. code-block:: yaml
+
+	myservice:
```
Comment (thatch45) [inline doc/ref/states/requisites.rst]: MOAR tabs

**Review by thatch45 — CHANGES_REQUESTED** [review]
This looks good, just need to fix those tabs :)

File: doc/ref/states/requisites.rst (hunk 1 above)
Comment (garethgreenaway) [reply doc/ref/states/requisites.rst]: Well this is awkward 😀

**Comment by garethgreenaway** [comment]
re-run ubuntu

File: doc/ref/states/requisites.rst [hunk 3]
```diff
@@ -245,6 +252,44 @@ This will add all of the state declarations found in the given sls file. This me
...
 large groups of states easily in any requisite statement.
 
+.. _requisites-require_any:
+
+require_any
+~~~~~~~~~~~
+
+.. versionadded:: Oxygen
+
+The use of ``require_any`` demands that one of the required state executes before the
```
Comment (rallytime) [inline doc/ref/states/requisites.rst]: `state` --> `states`

File: doc/ref/states/requisites.rst [hunk 4]
```diff
@@ -332,6 +377,50 @@ to Salt ensuring that the service is running.
...
+          - file: /etc/apache2/sites-available/site1.conf
+          - file: /etc/apache2/sites-available/site2.conf
+      file.managed:
+        - name: /etc/apache2/sites-available/site1.conf
+        - source: salt://apache2/files/site1.conf
+      file.managed:
+        - name: /etc/apache2/sites-available/site2.conf
+        - source: salt://apache2/files/site2.conf
+
+In this example, the service will be reloaded/restarted if the either of the state
```
Comment (rallytime) [inline doc/ref/states/requisites.rst]: I think this statement could be a bit more clear/specific. How about "In this example, the service will be reloaded/restarted if either of the ``file.managed`` states has a result of ``True`` and has changes."?

File: doc/ref/states/requisites.rst [hunk 5]
```diff
@@ -423,6 +512,43 @@ The ``onfail`` requisite is applied in the same way as ``require`` as ``watch``:
 
 .. _Issue #22370: https://github.com/saltstack/salt/issues/22370
 
+.. _requisites-onfail:
```
Comment (rallytime) [inline doc/ref/states/requisites.rst]: `_requisites-onfail` --> `_requisites-onfail_any`. (The previous wording is already used in this document.)

File: doc/ref/states/requisites.rst [hunk 6]
```diff
@@ -423,6 +512,43 @@ The ``onfail`` requisite is applied in the same way as ``require`` as ``watch``:
...
+~~~~~~~~~~
+
+.. versionadded:: Oxygen
+
+The ``onfail_any`` requisite allows for reactions to happen strictly as a response
+to the failure of at least one other state. This can be used in a number of ways, such as
+executing a second attempt to set up a service or begin to execute a separate
+thread of states because of a failure.
+
+The ``onfail_any`` requisite is applied in the same way as ``require_any`` as ``watch_any``:
```
Comment (rallytime) [inline doc/ref/states/requisites.rst]: The second `as`  should be `and`.

File: doc/ref/states/requisites.rst [hunk 7]
```diff
@@ -423,6 +512,43 @@ The ``onfail`` requisite is applied in the same way as ``require`` as ``watch``:
...
+        - name: /mnt/code
+        - device: 10.0.0.45:/code
+        - fstype: nfs
+
+    backup_mount:
+      mount.mounted:
+        - name: /mnt/share
+        - device: 192.168.40.34:/share
+        - fstype: nfs
+        - onfail:
```
Comment (rallytime) [inline doc/ref/states/requisites.rst]: `onfail` --> `onfail_any` :)

File: doc/ref/states/requisites.rst [hunk 8]
```diff
@@ -423,6 +512,43 @@ The ``onfail`` requisite is applied in the same way as ``require`` as ``watch``:
...
+          - mount: primary_mount
+          - mount: secondary_mount
+
```
Comment (rallytime) [inline doc/ref/states/requisites.rst]: Can you add a sentence here about what the expected outcome of the state run will be, as you've done for the other examples so far?

File: doc/ref/states/requisites.rst [hunk 9]
```diff
@@ -482,6 +608,37 @@ if any of the watched states changes.
...
+        - name: /etc/myservice/myservice.conf
+        - source: salt://myservice/files/myservice.conf
+        - mode: 600
+      file.managed:
+        - name: /etc/yourservice/yourservice.conf
+        - source: salt://yourservice/files/yourservice.conf
+        - mode: 600
+      cmd.run:
+        - name: /usr/libexec/myservice/post-changes-hook.sh
+        - onchanges:
```
Comment (rallytime) [inline doc/ref/states/requisites.rst]: `onchanges` --> `onchanges_any`

File: doc/ref/states/requisites.rst [hunk 10]
```diff
@@ -482,6 +608,37 @@ if any of the watched states changes.
...
+          - file: /etc/myservice/myservice.conf
+          - file: /etc/your_service/yourservice.conf
+
```
Comment (rallytime) [inline doc/ref/states/requisites.rst]: Same comment here about adding an example of what the expected state run will be.

**Review by rallytime — CHANGES_REQUESTED** [review]
Cool! And yay for tests!

I have some documentation changes that should be fixed up, though. :)

**Comment by rallytime** [comment]
Oh! And these new requisites should be added to the `Oxygen` release notes as well.

**Comment by garethgreenaway** [comment]
👍 will do another push to the PR tomorrow.

**Review by rallytime — APPROVED** [review]

**Review by thatch45 — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

**Pull request merged**